cryostatio/cryostat
# feat(hotreload): create configuration for hot-reload dev server

Fixes #744

To test:

1. In first terminal, `sh run-local.sh`. If you use `mvnd` or something else like I do and have this aliased as `mvn` in your `PATH`, then you may wish to do `MVN=/usr/bin/mvn sh run-local.sh` instead.
2. (Optional) In second terminal, `cd web-client && yarn yarn:frzinstall && yarn start:dev`

This should start a hot-reloading Cryostat backend as a local JVM process in the first terminal, and the typical existing hot-reloading Cryostat web-client in the second terminal. There will be no sample app targets and there will be no Grafana dashboard, but the Cryostat JVM should discover itself via JDP and you should be able to query it for its active recordings, event types, and event templates, as well as start and archive recordings.

Report generation doesn't currently work with this setup, so I'm looking into that. ~~It seems like it's probably a hostname resolution issue.~~
Turns out the report generation issue is actually a classpath problem - the main process isn't able to load the subprocess class to invoke:

```
Error: Could not find or load main class io.cryostat.net.reports.SubprocessReportGenerator
Caused by: java.lang.ClassNotFoundException: io.cryostat.net.reports.SubprocessReportGenerator
Nov. 03, 2021 3:31:35 P.M. io.cryostat.core.log.Logger error
SEVERE: Exception thrown
io.cryostat.net.reports.SubprocessReportGenerator$ReportGenerationException: [1] Connection to target JVM failed.
...
```

This has two parts. First, the new `vertx-maven-plugin` configuration added doesn't have any easy way that I can see to simply set/append to the classpath when using `vertx:run`. It may be cleared up by using `vertx:start` (https://reactiverse.io/vertx-maven-plugin/#vertx:start) instead. Second, the way we invoke the subprocess uses a hardcoded classpath (https://github.com/cryostatio/cryostat/blob/460e2fd78646965cc0759b6431526e5b2c1cac73/src/main/java/io/cryostat/util/JavaProcess.java#L60), which is very unlikely to actually match the filesystem on a development host machine. We would need to make this configurable by an env var or system property so that it can be set to point to the maven build `target` directory contents.
